File: backend/internal/service/coach_service.go
package service

import (
	"errors"
	"fmt"
	"gym-admin/internal/models"
	"gym-admin/internal/repository"
	"time"
)

type CoachService struct {
	repo *repository.CoachRepository
}

func NewCoachService() *CoachService {
	return &CoachService{
		repo: repository.NewCoachRepository(),
	}
}

func (s *CoachService) CreateCoach(coach *models.Coach) error {
	// Generate coach number
	coach.CoachNo = s.generateCoachNo()
	coach.Status = 1 // Default status: active

	return s.repo.Create(coach)
}

func (s *CoachService) GetCoach(id int64) (*models.Coach, error) {
	return s.repo.GetByID(id)
}

func (s *CoachService) ListCoaches(page, pageSize int, status *int8) ([]models.Coach, int64, error) {
	if page < 1 {
		page = 1
	}
	if pageSize < 1 || pageSize > 100 {
		pageSize = 10
	}
	return s.repo.List(page, pageSize, status)
}

func (s *CoachService) UpdateCoach(id int64, updates map[string]interface{}) error {
	coach, err := s.repo.GetByID(id)
	if err != nil {
		return errors.New("coach not found")
	}

	// Update fields
	if name, ok := updates["name"].(string); ok {
		coach.Name = name
	}
	if phone, ok := updates["phone"].(string); ok {
		coach.Phone = phone
	}
	if email, ok := updates["email"].(string); ok {
		coach.Email = email
	}
	if introduction, ok := updates["introduction"].(string); ok {
		coach.Introduction = introduction
	}

	return s.repo.Update(coach)
}

func (s *CoachService) DeleteCoach(id int64) error {
	return s.repo.Delete(id)
}

func (s *CoachService) generateCoachNo() string {
	return fmt.Sprintf("C%d", time.Now().UnixNano()/1000000)
}

func (s *CoachService) GetAllCoaches() ([]models.Coach, error) {
	return s.repo.GetAll()
}

File: frontend/src/services/coachService.ts
import apiClient from './api'

export interface Coach {
  id: number
  coach_no: string
  name: string
  phone: string
  gender?: number
  email?: string
  specialties?: string
  experience?: number
  status: number
  created_at: string
}

export interface CoachListResponse {
  code: number
  message: string
  data: {
    list: Coach[]
    total: number
    page: number
    page_size: number
  }
}

export const coachService = {
  list: async (params?: any): Promise<CoachListResponse> => {
    return apiClient.get('/coaches', { params })
  },

  get: async (id: number): Promise<Coach> => {
    return apiClient.get(`/coaches/${id}`)
  },

  create: async (data: Partial<Coach>): Promise<Coach> => {
    return apiClient.post('/coaches', data)
  },

  update: async (id: number, data: Partial<Coach>): Promise<any> => {
    return apiClient.put(`/coaches/${id}`, data)
  },

  delete: async (id: number): Promise<any> => {
    return apiClient.delete(`/coaches/${id}`)
  },
}
